### PR TITLE
SoundSourceM4A: Reconfigure decoder after channel config errors

### PR DESCRIPTION
**Pull request by uklotzde**
A fix for the evil FAAD2 2.9.x that required a huge amount of work and testing. Continued and finished after recent discussions in #2738.

I still don't understand why we need all those complex workarounds to decode regular M4A files. VLC is also using lots of custom patches to fix issues.

If more issues arise I recommend to drop FAAD2 entirely. It's not worth to waste so much time on this.

### CODE REVIEW

**Comment by Holzhaus** [comment]
Any reasons why we didn't drop FAAD when FFmpeg support was added?

**Comment by uklotzde** [comment]
Offsets for cue points might differ slightly.

On the other hand it would allow us to drop libmp4v2 which also seems to be discontinued.

**Comment by Holzhaus** [comment]
The offsets would be the same across all OSes, right? If we find a pattern we can correct them with a DB migration. We should switch eventually IMHO.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you for the fix and all the useful refactoring. 

I am unsure how we should continue here in the long run. Since libmp4v2 was removed from debian. 
The faad guys have fixed it by adding their own mp4 stream extraction functions. 

During review I have done some fixes
https://github.com/uklotzde/mixxx/pull/15

**Comment by uklotzde** [comment]
Ping. This is needed for fixing .m4a playback on Fedora 32, replacing 2.2 with 2.3 beta. Otherwise I have to do a switch from FAAD2 to FFmpeg permanently.

File: src/sources/soundsourcem4a.cpp [hunk 1]
```diff
@@ -372,213 +510,308 @@ ReadableSampleFrames SoundSourceM4A::readSampleFramesClamped(
...
+                    m_inputBufferLength - m_inputBufferOffset,
+                    reinterpret_cast<void**>(&pDecodeBuffer),
+                    decodeBufferCapacity * sizeof(*pDecodeBuffer));
+            if (decFrameInfo.error != 0) {
+                // A decoding error has occurred
+                if (retryAfterReopeningDecoder) {
+                    // At this point we have failed to decode the current sample
+                    // block twice and need to discard it. The content of the
+                    // sample block is unknown and we simply continue with the
+                    // next block. This is just a workaround! The reason why FAAD2
```
Comment (daschuer) [inline src/sources/soundsourcem4a.cpp:685]: by using silence instead  ..
or 
by skipping the track. 

I am afraid that the later will produce a cue point offset if the track was originally decoded with a working FAAD2 decoder or if it was decoded with a broken decoder the offset will appear after the decoder is fixed.

**Review by daschuer — CHANGES_REQUESTED** [review]

File: src/sources/soundsourcem4a.cpp (hunk 1 above)
Comment (uklotzde) [reply src/sources/soundsourcem4a.cpp:685]: I am not writing an AAC decoder for the unknown frames. No chance. In this case I will close this PR.

Comment (uklotzde) [reply src/sources/soundsourcem4a.cpp:685]: We don't even know if and how many samples those blocks contain. Sorry, but this is not doable.

**Comment by daschuer** [comment]
No, need to fix that. I was just curious to know what this patch means.
Please add a comment about the offset and we can merge.

**Comment by uklotzde** [comment]
*sigh*

**Review by daschuer — APPROVED** [review]
Thank you. LGTM

**Pull request merged**